src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp

```diff
@@ -292,21 +316,25 @@ struct RemoveRecursiveObjectStorageOperation final : public IDiskObjectStorageOp
     const bool keep_all_batch_data;
     /// paths inside the 'this->path'
     const NameSet file_names_remove_metadata_only;
+    const UInt64 batch_size;
 
-    /// map from local_path to its remote objects with hardlinks counter
+    /// local_path and related remote objects with hardlinks counter
     /// local_path is the path inside 'this->path'
-    std::unordered_map<std::string, ObjectsToRemove> objects_to_remove_by_path;
+    std::vector<std::string> local_paths_to_remove;
+    std::vector<ObjectsToRemove> objects_to_remove;
 
     RemoveRecursiveObjectStorageOperation(
         IObjectStorage & object_storage_,
         IMetadataStorage & metadata_storage_,
         const std::string & path_,
         bool keep_all_batch_data_,
-        const NameSet & file_names_remove_metadata_only_)
+        const NameSet & file_names_remove_metadata_only_,
+        UInt64 batch_size_)
```

Comment:
Imho `batch_size` is not semantically suitable for the operation, I propose to call it a `limit`